---
title: "R Notebook"
output: html_notebook
---

```{r}
library("airway")
library("Rsamtools")
library("GenomicFeatures")
library("GenomicAlignments")
library("BiocParallel")
library("DESeq2")
library("tidyverse")
library("ggsci")
theme_set(theme_classic())
in_dir <- system.file("extdata", package = "airway", mustWork = TRUE)
```

# sample table
```{r}
csv_file <- file.path(in_dir, "sample_table.csv")
sample_table <- read.csv(csv_file, row.names = 1)
```

# alignment
```{r data}
file_names <- file.path(in_dir, paste0(sample_table$Run, "_subset.bam"))
bam_files <- BamFileList(file_names, yieldSize = 2000000)
```

# annotation
```{r}
gtf_file <- file.path(in_dir, "Homo_sapiens.GRCh37.75_subset.gtf")
txdb <- makeTxDbFromGFF(gtf_file, format = "gtf", circ_seqs = character())
txdb
```

```{r}
ebg <- exonsBy(txdb, by="gene")
ebg
```

# Counts
```{r}
se <- summarizeOverlaps(features = ebg, reads = bam_files,
                        mode = "Union",
                        singleEnd = FALSE,
                        ignore.strand = TRUE,
                        fragments = TRUE )
```

```{r}
dir.create("data")
lapply(colnames(assay(se)), function(x) {
  sample_name <- gsub("_subset.bam", "", x)
  assay(se) %>%
    as.data.frame() %>%
    dplyr::select(sample_name = x) %>% 
    rownames_to_column("gene") %>%
    write_tsv(., paste0("data/", sample_name, ".txt"))
})
```

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = col_data,
                              design = ~ cell + dex)
```

